### ISSUE TITLE
after_commit callbacks within transitions

### ISSUE THREAD

**Issue by take-five** [aasm/aasm#224]
Hi,

I want to trigger different callbacks on certain transitions within one event.

```ruby
aasm do
  event :prolong do
    transitions from: :awaiting, to: :active do
     after { NotificationWorker.perform_async(id, :activated) }
    end

    transitions from: :active, to: :active do
      after { NotificationWorker.perform_async(id, :prolonged) }
    end

    transitions from: :suspended, to: :active do
      after { NotificationWorker.perform_async(id, :prolonged) }
    end
  end
end
```

It works like a charm, but the problem here is that these callbacks get fired right after transition, even before any changes got persisted in DB. In my case callbacks enqueue Sidekiq jobs and cannot be rolled back. If any error occur during persisting record in DB, these jobs would be enqueued anyway, and that behaviour can break consistency.

Is there any way to add `after_commit` callbacks to AASM?

**Comment by alto** [comment]
@take-five 

There are `after_commit` callbacks on events, which you can use to do something like this:

```ruby
  aasm do
    event :prolong, after_commit: :notify do
      transitions from: :awaiting,  to: :active
      transitions from: :active,    to: :active
      transitions from: :suspended, to: :active
    end
  end

  def notify
    case [aasm.from_state, aasm.to_state]
    when [:awaiting,  :active]
      NotificationWorker.perform_async(id, :activated)
    when [:active,    :active]
      NotificationWorker.perform_async(id, :prolonged)
    when [:suspended, :active]
      NotificationWorker.perform_async(id, :prolonged)
    end
  end
```

This could be written more concisely or nicely, but you get the idea.

Does that help?

**Comment by hpoydar** [comment]
So it's safe to say that putting jobs into a queue that depend on the persisted end state of the object need to be in an after_commit callback, no?

Can the DSL handle blocks for both the after and after_commit callbacks like this?  

    aasm do

      state :created
      state :active

      event :activate
        after_commit do
          do_something_job_will_need!
          Job.perform_later(self)
        end
        after do
          log_state_change
        end
        transitions from: :created, to: :active
      end

    end

**Comment by alto** [comment]
@hpoydar Yes, please use `after_commit` event callbacks for that.

I just released version `4.1.1`, including a patch with which the block notation for `after_commit` callbacks is supported all (`after` was supported before). So, both callbacks support the block notation now.

Make sure to add a `do` at the end of line `event :activate` though!

**Comment by alto** [comment]
@take-five I assume your problem is solved. Closing this issue now.
